Feat: Migrated the infrastructure utility files from the ts-core project

jcdubs-leighton opened the pull request:
## Summary

* Migrated the generate resource name utility function from ts-core.
* Migrated the generate s3 bucket name utility function from ts-core.
* Migrated the get removal policy from stage utility function from ts-core.
* Migrated the get stage utility function from ts-core.
* Updated the docs structure to include the utility function directory.
* Updated the docs to have a side bar with consistent naming by adding READMEs to the top level directories.

## Checklist
- [x] No secrets or credentials added
- [x] Updated SECURITY.md if needed
- [x] Updated documentation if needed

ro2nie commented on src/types/README.md:
```diff
@@ -0,0 +1,54 @@
+The types in this directory serve to:
+
+- **Ensure type safety** across all constructs, utilities, and consuming applications
+- **Standardize common patterns** like resource naming, tagging, and environment configuration
```
🇬🇧 standardise?

g-farrow commented on src/utils/generate-resource-name/generate-resource-name.ts:
```diff
@@ -0,0 +1,57 @@
+ * // "dev-cache-bucket-temp"
+ */
+export function generateResourceName(
+  stage: ResourceNameParts['stage'],
+  service: ResourceNameParts['service'],
+  resource: ResourceNameParts['resource'],
+  suffix?: ResourceNameParts['suffix'],
+  region?: ResourceNameParts['region'],
```
positional arguments can be problematic. its easy to accidentally pass arguments in the wrong order, whereas with named arguments, this cannot be done. Also, where there are optional arguments, having to place pointless "undefined" entries when calling it is not great.
I think we should swap this to be named arguments.
Also, we could just use the type provided by resource-name-types. something like:
```
{ stage, service, resource, suffix, region}: ResourceNameParts
```

ro2nie replied:
positional arguments here make it easy to get them mixed up. If we were using a single typed object, ordering wouldn't matter, and it would get rid of undefined ones when not wanting to use one.

g-farrow commented on src/utils/generate-s3-bucket-name/generate-s3-bucket-name.ts:
```diff
@@ -0,0 +1,73 @@
+ * // "staging-analytics-bucket-logs"
+ */
+export function generateS3BucketName(
+  stage: ResourceNameParts['stage'],
+  service: ResourceNameParts['service'],
+  suffix?: ResourceNameParts['suffix'],
+  region?: ResourceNameParts['region'],
```
same comment about positional vs named arguments

g-farrow commented on src/utils/generate-s3-bucket-name/generate-s3-bucket-name.ts:
```diff
@@ -0,0 +1,73 @@
+    .join('-')
+    .toLowerCase();
+
+  if (name.length < MIN_LENGTH) {
+    throw new Error(
+      `Generated S3 bucket name "${name}" is too short. Must be at least ${MIN_LENGTH} characters.`,
+    );
+  }
+
+  if (name.length > MAX_LENGTH) {
+    throw new Error(
+      `Generated S3 bucket name "${name}" exceeds the maximum allowed length of ${MAX_LENGTH} characters.`,
+    );
+  }
+
+  if (!/^[a-z0-9][a-z0-9.-]*[a-z0-9]$/.test(name)) {
+    throw new Error(
+      `Generated S3 bucket name "${name}" contains invalid characters. Must contain only lowercase letters, numbers, dots, and hyphens, and start/end with alphanumeric characters.`,
+    );
+  }
```
as above: we already have a dependency on zod, so we could use zod schemas for this type of thing to clean it up. they can be reusable

g-farrow commented on src/utils/generate-resource-name/generate-resource-name.ts:
```diff
@@ -0,0 +1,57 @@
+    .join('-')
+    .toLowerCase();
+
+  if (name.length > MAX_LENGTH) {
+    throw new Error(
+      `Generated resource name "${name}" exceeds the maximum allowed length of ${MAX_LENGTH} characters.`,
+    );
+  }
```
we already have a dependency on zod, so we could use zod schemas for this type of thing to clean it up. they can be reusable
also, is this missing some checks on invalid characters etc?

g-farrow commented on src/utils/get-stage/get-stage.ts:
```diff
@@ -0,0 +1,27 @@
+ */
+
+export function getStage(stage: string): string {
+  switch (stage) {
+    case Stage.prod:
+      return Stage.prod;
+    case Stage.staging:
+      return Stage.staging;
+    case Stage.test:
+      return Stage.test;
+    default:
+      return stage.toLowerCase(); // return the ephemeral environment if not known (i.e. pr-123)
+  }
```
this is similar to the switch we had in the DNS utils in ts-core. It can be replaced with a simpler implementation that doesn't use switch:

ro2nie commented on src/utils/get-removal-policy-from-stage/get-removal-policy-from-stage.ts:
```diff
@@ -0,0 +1,32 @@
+ */
+export function getRemovalPolicyFromStage(stage: string): RemovalPolicy {
+  if (
+    stage.toLowerCase().trim() !== Stage.prod &&
+    stage.toLowerCase().trim() !== Stage.staging
+  ) {
```
[Stage.prod, Stage.staging].includes(stage.toLowerCase().trim())

ro2nie reviewed (changes requested):
Left some comments

jcdubs-leighton commented:
@g-farrow and @ro2nie I've addressed all comments and pushed up changes apart from the following comment:

https://github.com/leighton-digital/cloud-blocks/pull/12#discussion_r2405677128

We will discuss this in the meeting this afternoon but everything else is ready for review again.

jcdubs-leighton commented on src/utils/README.md:
````diff
@@ -0,0 +1,434 @@
+### ResourceNameParts Interface
+
+```ts
+interface ResourceNameParts {
+  /** Environment stage (e.g., 'dev', 'prod') */
+  stage: string;
+
+  /** Service or domain name (e.g., 'orders', 'users') */
+  service: string;
+
+  /** Resource type (e.g., 'table', 'queue', 'function') */
+  resource: string;
+
+  /** Optional suffix for uniqueness (e.g., 'v2', 'backup') */
+  suffix?: string;
+
+  /** Optional AWS region (e.g., 'us-east-1') */
+  region?: Region;
+}
````
We are going to leave this logic in based on our discussion in the OSS meeting yesterday.

DanielMackenzie93 commented on src/types/environments.ts:
```diff
@@ -0,0 +1,48 @@
+  develop = 'develop',
+  staging = 'staging',
+  prod = 'prod',
+  test = 'test',
```
**question**: Did @leegilmorecode mention yesterday that test wasn't a thing anymore? (Apologies, not aware of context around this, just incase something has been missed)

g-farrow commented on package.json:
```diff
@@ -72,6 +74,8 @@
     "@commitlint/cli": "^19.8.1",
     "@commitlint/config-conventional": "^19.8.1",
+    "@cspell/dict-en-gb": "^5.0.14",
+    "@cspell/eslint-plugin": "^9.2.1",
```
eslint plugin? is this correct?

g-farrow commented on src/types/environments.ts:
```diff
@@ -0,0 +1,48 @@
+ * @enum {string}
+ */
+export enum Region {
+  dublin = 'eu-west-1',
+  london = 'eu-west-2',
+  frankfurt = 'eu-central-1',
+  virginia = 'us-east-1',
+  ohio = 'us-east-2',
+  oregon = 'us-west-2',
+  california = 'us-west-1',
+  canada = 'ca-central-1',
+  saoPaulo = 'sa-east-1',
+  stockholm = 'eu-north-1',
+  milan = 'eu-south-1',
+  paris = 'eu-west-3',
+  ireland = 'eu-west-1',
+  capeTown = 'af-south-1',
+  bahrain = 'me-south-1',
+  singapore = 'ap-southeast-1',
+  sydney = 'ap-southeast-2',
+  jakarta = 'ap-southeast-3',
+  tokyo = 'ap-northeast-1',
+  seoul = 'ap-northeast-2',
+  osaka = 'ap-northeast-3',
+  mumbai = 'ap-south-1',
+  hyderabad = 'ap-south-2',
+  hongKong = 'ap-east-1',
+  uae = 'me-central-1',
+  zurich = 'eu-central-2',
+  spain = 'eu-south-2',
+  melbourne = 'ap-southeast-4',
+  israel = 'il-central-1',
```
Enums usually use PascalCase in Typescript as a convention, e.g. `Region.Frankfurt`, I think we should stick to this unless we have a standard/convention to suggest otherwise.

Same goes for the other enums in this file

ro2nie commented on src/utils/generate-s3-bucket-name/generate-s3-bucket-name.ts:
```diff
@@ -1,8 +1,18 @@
+  .min(3, 'is too short. Must be at least 3 characters.')
+  .max(63, 'exceeds the maximum allowed length of 63 characters.')
+  .regex(
+    /^[a-z0-9][a-z0-9.-]*[a-z0-9]$/,
```
{1, 64}

ro2nie replied:
nit: perhaps add quantifier? {1,64}

g-farrow commented on src/utils/get-stage/get-stage.ts:
```diff
@@ -0,0 +1,8 @@
+ * @returns The normalized stage string.
+ */
+export function getStage(stage: string): string {
+  return stage.toLowerCase();
```
could it be necessary for additional sanitising hear? e.g. `.trim()`

ro2nie commented on src/utils/README.md:
```diff
@@ -0,0 +1,434 @@
+
+A collection of **AWS CDK infrastructure utilities** for **consistent resource naming and environment management**, featuring:
+
+* **Standardized resource naming** with `generateResourceName` and `generateS3BucketName`
```
🇬🇧 spelling

ro2nie commented on src/utils/README.md:
```diff
@@ -0,0 +1,434 @@
+    super(scope, id, props);
+
+    const { stage, service } = props;
+    const normalizedStage = getStage(stage);
```
🇬🇧 spelling

ro2nie commented on src/utils/README.md:
```diff
@@ -0,0 +1,434 @@
+* **Global uniqueness**: S3 bucket names must be globally unique across all AWS accounts
+* **Character restrictions**: S3 buckets cannot contain uppercase letters or underscores
+* **Length limits**: Function validates against AWS service limits during synthesis
+* **Stage normalization**: Unknown stages are converted to lowercase for consistency
```
🇬🇧 spelling

ro2nie commented on src/utils/get-stage/get-stage.ts:
```diff
@@ -0,0 +1,8 @@
+/**
+ * Normalizes the stage string to lowercase to ensure consistency.
```
🇬🇧 spelling
